FlintCore/Utils/CompletionRequirement.swift

```diff
@@ -0,0 +1,210 @@
+                }
+                return result
+            }
+        }
+        
+        init() {
+            _value = nil
+        }
+
+        init(result: T) {
+            _value = result
+        }
+        
+        public var isCompletingAsync: Bool { return false }
+    }
+
+    // The type for a status indicating completion will occur later
+    public class DeferredStatus: Status {
+        var owner: CompletionRequirement<T>?
+        
+        init(owner: CompletionRequirement<T>) {
+            self.owner = owner
+            super.init()
+        }
+
+        override public var isCompletingAsync: Bool {
+            return true
+        }
+        
+        public func completed(_ result: T) {
+            guard let owner = owner else {
+                return
+            }
+            owner.completionHandler(result, true)
+        }
+    }
+
+    fileprivate var completionStatus: Status?
+    var completionHandler: ((T, _ completedAsync: Bool) -> Void)!
+
+    public init(completionHandler: @escaping (T, _ completedAsync: Bool) -> Void) {
+        self.completionHandler = completionHandler
+    }
+
+    /// Internal initialiser for proxy subclass, which will set the completion after
+    fileprivate init() {
+    }
+
+    /// Call to verify that the result belongs to this completion instance and there hasn't been a mistake
+    public func verify(_ status: Status) -> Bool {
+        return status === completionStatus
+    }
+    
+    /// Call to indicate that completion will be called later, asynchronously by code that has a reference
+    /// to the deferred status.
+    public func willCompleteAsync() -> DeferredStatus {
+        guard completionStatus == nil else {
+            flintUsageError("Only one of completedSync() or willCompleteLater() can be called")
+        }
+        
+        // Set "async" execution
+        let status = DeferredStatus(owner: self)
+        completionStatus = status
+        
+        // Return a status to inform the caller
+        return status
+    }
+
+    /// Call to indicate that completion is to be called immediately, synchronously
+    public func completedSync(_ result: T) -> Status {
+        guard self.completionStatus == nil else {
+            flintUsageError("Only one of completedSync() or willCompleteLater() can be called")
+        }
+        
+        let completionStatus = Status(result: result)
+        completionHandler(result, false)
+        self.completionStatus = completionStatus
+        return completionStatus
+    }
+}
+
+/// A `ProxyCompletionRequirement` allows you to provide a completion requirement that adds some custom completion logic
+/// to an existing completion instance, and then return a possibly modified result value to the original requirement.
+///
+/// Very much turtles all the way down, and a bit nasty in the nuance of the implementation.
+public class ProxyCompletionRequirement<T>: CompletionRequirement<T> {
```

Comment:
Would it be easier or nicer to have a queue and just be able to put "proxy completions" on it instead of wrapping it? I could imagine it getting complicated if you have multiple proxies that you try to put together. I've seen code like that get out of hand.

Reply:
Yes, I think this definitely gets complex, although it has been pretty easy to work with so far. It is really meant for internal use only, but it will sometimes be needed by app Action(s) that need to call other actions. However I'm not sure of a better solution at the moment, because the semantics are really tricky. I know that this is going to cause complications later, but it is hard to foresee what these are without using it in anger.